# Scripts or tools used in the analysis

## Introduction
The repository deposits all customized R/Python/BASH code used for the analysis by study *xxx et. al.* (DOI: xxx) for the purpose of reproduction.

## Table of tools/packages used in the analysis

| Tool name | Version | Comments |
| :-------: | :-----: | :------: |
| PLINK     |         |          |
| BCFtools  |         |          |
| TODO      |         |          |


## Statistical analysis
The statistical analyses were performed by in-house R/Python script or published tools/packages.
The codes used to performed each analysis are listed and briefly introduced as the following.

### Single-cell RNA-seq analysis and cell type identification
1. Demultiplexing by `souporcell`
2. Single-cell analysis by `Seurat`
3. Cell type identification by `?`

### Identification of co-expression modules

We used weighted gene co-expression network analysis (WGCNA) to detect co-expression modules for each transcriptomic profiles of four identified cell types.
For more details about the analysis check R code at `r/wgcna.r`

### Definition of pseudo-time windows
`r/pseudotime.r` **TODO**

### Identification of single-cell expression quantitative traits loci (sceQTL)
`snakemake/300bcg_limix_qtl.smk`

### Estimation of shared signals
`r/mashr.r`

### Construction of reference genotype panel
To speed up the calculation of harmonization step, we prepared a reference panel using genetic variants from 1000 genome project.
The generated reference panel was also used to estimate enrichement of the identified QTL SNPs in epigenomic annotations.
The detailed steps is recorded in `bash/prepare_reference.sh`

### Harmonization of genetic variants
`r/harmonization.r`

### Mendelian randomization alaysis
`r/mr.r`

### Colocalization analysis
`r/coloc.r`

### Enrichement of identified QTL SNPs in epigenomic annotations
We obtained celltype-specific ATAC-seq peaks from three cell types, namely monocytes, CD8T, and NK cells, sorted from PBMC samples at d0 ("V1") in quantification*.csv.gz files.
`bash/enrich.sh` and `r/enrich.r`.


## Publicly available data and reference

### Table of publicly available GWAS summary statistics

### Table of reference data

| Reference name | Version | Availability | Comments |
| :------------: | :-----: | :----------: | :------: |
| Human genome   | GRCh38  |              |          |
| Genetic variants | GRCh38/build xxx | xxx | |


## Misc

SingleR reference

| Database                            | Species | Comments                                                                                                  |
| :---------------------------------: | :-----: | :-------------------------------------------------------------------------------------------------------: |
| Monaco immnune data                 | Human   | RNA-Seq profiling on 29 immune cell types consituting PBMCs sorted from 4 Singaporean-Chinese individuals |
| Human Primary Cell Atlas            | Human   | Microarray datasets derived from human primary cells                                                      |
| Blueprint / ENCODE                  | Human   | Bulk RNA-seq data for pure stroma and immune cells generated by Blueprint                                 |
| Database of Immnune Cell Expression | Human   | Bulk RNA-seq samples of sorted human immune cell populations                                              |
